### PR TITLE
[Buttons] Add state-based border color and width

### PR DESCRIPTION
**Pull request by samnm**

### CODE REVIEW

File: components/Buttons/examples/ButtonsTypicalUse.m [hunk 1]
```diff
@@ -30,9 +30,11 @@ - (MDCButton *)buildCustomStrokedButton {
   [button setBackgroundColor:[UIColor clearColor] forState:UIControlStateNormal];
   [button setTitleColor:[UIColor colorWithWhite:0.1f alpha:1] forState:UIControlStateNormal];
   button.inkColor = [UIColor colorWithWhite:0 alpha:0.06f];
-  button.layer.borderWidth = 1;
-  button.layer.borderColor = [UIColor blackColor].CGColor;
   button.disabledAlpha = 0.38f;
+
+  [button setBorderWidth:1.0 forState:UIControlStateNormal];
+  [button setBorderColor:[UIColor blackColor] forState:UIControlStateNormal];
```
Comment (ianegordon) [inline components/Buttons/examples/ButtonsTypicalUse.m]: Should stroke color be the same color as Title Color?
[UIColor colorWithWhite:0.1f alpha:1]

File: components/Buttons/src/MDCButton.m [hunk 2]
```diff
@@ -550,6 +578,56 @@ - (void)setElevation:(CGFloat)elevation forState:(UIControlState)state {
...
 }
 
+#pragma mark - Border Color
+
+- (UIColor *)borderColorForState:(UIControlState)state {
+  return _borderColors[@(state)];
+}
+
+- (void)setBorderColor:(UIColor *)borderColor forState:(UIControlState)state {
+  if (!_borderColors) {
```
Comment (ianegordon) [inline components/Buttons/src/MDCButton.m]: Isn't _borderColors created in commonInit?
Redundant?

File: components/Buttons/src/MDCButton.m [hunk 3]
```diff
@@ -550,6 +578,56 @@ - (void)setElevation:(CGFloat)elevation forState:(UIControlState)state {
...
+- (CGFloat)borderWidthForState:(UIControlState)state {
+  NSNumber *borderWidth = _borderWidths[@(state)];
+  if (borderWidth) {
+    return borderWidth.doubleValue;
+  }
+  return 0;
+}
+
+- (void)setBorderWidth:(CGFloat)borderWidth forState:(UIControlState)state {
+  if (!_borderWidths) {
```
Comment (ianegordon) [inline components/Buttons/src/MDCButton.m]: Isn't borderWidths created in commonInit?
Redundant?

File: components/Buttons/src/MDCButton.m [hunk 4]
```diff
@@ -550,6 +578,56 @@ - (void)setElevation:(CGFloat)elevation forState:(UIControlState)state {
...
+    _borderColors = [NSMutableDictionary dictionary];
+  }
+  _borderColors[@(state)] = borderColor;
+
+  [self updateBorderColor];
+}
+
+- (void)updateBorderColor {
+  UIColor *color = _borderColors[@(self.state)];
+  if (!color && self.state != UIControlStateNormal) {
```
Comment (ianegordon) [inline components/Buttons/src/MDCButton.m:612]: Consider adding a comment here and in updateBorderWidth
// Use value for UIControllerStateNormal if self.state isn't defined

**Review by ianegordon — CHANGES_REQUESTED** [review]

**Review by ianegordon — COMMENTED** [review]
Do we need to update the associated button themers as well?

**Comment by samnm** [comment]
Fixed reviewed issues.

There are no button themers at the moment.

**Review by ianegordon — APPROVED** [review]

**Comment by ianegordon** [comment]
LGTM, please merge once Travis gives a +1.

**Pull request merged**